google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/UploadSourceDialog.java

```diff
@@ -73,18 +82,57 @@ public CredentialedUser getCredentialedUser() {
 
   @Override
   protected String getDimensionServiceKey() {
-    return "ChooseProjectDialog";
+    return "UploadSourceDialog";
   }
 
   private void createUIComponents() {
     projectSelector = new ProjectSelector();
     projectSelector.setMinimumSize(new Dimension(300, 0));
+
+    repositorySelector = new RepositorySelector(projectSelector.getText(),
+        projectSelector.getSelectedUser(), true /*canCreateRepository*/);
+
     projectSelector.getDocument().addDocumentListener(new DocumentAdapter() {
       @Override
       protected void textChanged(DocumentEvent event) {
-        setOKActionEnabled(projectSelector.getSelectedUser() != null);
+        repositorySelector.setCloudProject(projectSelector.getText());
+        repositorySelector.setUser(projectSelector.getSelectedUser());
+        repositorySelector.setText("");
+        updateButtons();
       }
     });
+
+    repositorySelector.getDocument().addDocumentListener(new DocumentAdapter() {
+      @Override
+      protected void textChanged(DocumentEvent e) {
+        updateButtons();
+      }
+    });
+
+    // todo figure out how to disable it when no project is selected
+//    repositorySelector.setEnabled(false);
+  }
+
+  private void updateButtons() {
+    if (!StringUtil.isEmpty(projectSelector.getText())
+        && projectSelector.getSelectedUser() == null) {
+      setErrorText("Invalid Cloud Project selected.");
+      setOKActionEnabled(false);
+      return;
+    } else if (!StringUtil.isEmpty(repositorySelector.getText())
+        && StringUtil.isEmpty(repositorySelector.getSelectedRepository())) {
+      setErrorText("Invalid Cloud Repository selected.");
+      setOKActionEnabled(false);
+      return;
+    } else if(projectSelector.getSelectedUser() == null
+        || StringUtil.isEmpty(repositorySelector.getSelectedRepository())) {
+      setErrorText(null);
+      setOKActionEnabled(false);
+      return;
+    }
```

Comment:
Part of it is the same. I think I can collapse the validation into the selector classes so the code can be reused.